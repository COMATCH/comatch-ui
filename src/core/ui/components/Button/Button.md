
## Button

From [`Button.js`](Button.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**className** | `String` | `''` | :x: | 
**disabled** | `Boolean` | `false` | :x: | 
**ghost** | `Boolean` | `false` | :x: | 
**href** | `String` | `''` | :x: | 
**icon** | `ReactNode` | `null` | :x: | 
**iconAfterText** | `Boolean` | `false` | :x: | 
**id** | `String` | `null` | :x: | 
**onClick** | `Function` | `noop` | :x: | 
**target** | `String` | `null` | :x: | 
**text** | `String` | `''` | :x: | 
**type** | `Enum('button','submit','reset')` | `'button'` | :x: | 



