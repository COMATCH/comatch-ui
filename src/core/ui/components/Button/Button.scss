@import '../../../styles/variables.scss';

.Button {
    border: 1px solid;
    border-radius: 3px;
    display: inline-block;
    font-size: $co-font-sm;
    outline: 0 !important;
    font-size: $co-font-sm;
    touch-action: manipulation;
    cursor: pointer;
    text-align: center;
    text-transform: uppercase;
    white-space: nowrap;
    padding: 8px 15px; // makes the height 40px
    transition: background-color 250ms ease-out, border-color 250ms ease-out, color 250ms ease-out;

    &.full {
        background-color: $co-color-primary;
        border-color: $co-color-primary;
        color: $co-color-white;

        &.disabled {
            background-color: $co-color-light-gray !important;
        }

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
        }
    }

    &.ghost {
        background-color: $co-color-white;
        border-color: $co-color-primary;
        color: $co-color-primary;

        &:hover {
            background-color: $co-color-secondary;
            border-color: $co-color-secondary;
            color: $co-color-white;
        }

        &.disabled {
            background-color: $co-color-white !important;
        }
    }

    &.disabled {
        border-color: $co-color-mid-gray !important;
        cursor: not-allowed;
        pointer-events: none;

        &,
        span {
            color: $co-color-mid-gray !important;
        }
    }

    // In case a Button has an icon defined
    svg {
        margin-right: 5px;
    }

    &.only-icon {
        svg {
            margin: 0;
        }
        padding: 8px 12px;
    }

    &.icon-after-text {
        white-space: initial;

        svg {
            float: right;
            position: relative;
            top: 2px;
            margin-left: 5px;
            margin-right: 0;
        }
    }

    // If buttons follow each other in a row
    // set a standardized whitespace between them
    & ~ & {
        margin-left: 9px;
    }
}
