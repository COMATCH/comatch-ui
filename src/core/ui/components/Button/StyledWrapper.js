import React from 'react';
import styled from 'styled-components';
import { palette, typography } from '../../../styles/variables';

export const generatePopupPositionBasedStyling = ({ popupMenuPosition }) => {
    switch (popupMenuPosition) {
        case 'bottom':
            return `
                left: 0;
                width: 100%;
                top: 42px;
            `;

        case 'top':
            return `
                left: 0;
                width: 100%;
                bottom: 42px;
            `;

        case 'left':
            return `
                top: -1px;
                right: 100%;
                margin-right: 5px;
            `;

        case 'right':
            return `
                top: -1px;
                left: 100%;
                margin-left: 5px;
            `;

        default:
            return '';
    }
};

export const generateSVGStyling = ({ iconAfterText, onlyIcon }) => `
    // Some .Button adjustments
    ${onlyIcon ? 'padding: 8px 12px;' : ''}

    svg {
        ${onlyIcon ? 'margin: 0;' : 'margin-right: 5px'}

        ${iconAfterText ? `
            float: right;
            position: relative;
            top: 2px;
            margin-left: 5px;
            margin-right: 0;
        ` : ''}
    }
`;

export const generateFullPropsStyling = ({ disabled, full }) =>
    full
        ? `
            background-color: ${disabled ? `${palette.lightGray} !important` : palette.primary};
            border-color: ${disabled ? `${palette.midGray} !important` : palette.primary};
            color: ${palette.white};

            &:hover {
                background-color: ${palette.secondary};
                border-color: ${palette.secondary};
            }
        `
        : '';

export const generateGhostStyling = ({ disabled, ghost, textOnly }) =>
    ghost || textOnly
        ? `
            background-color: ${disabled ? `${palette.white} !important` : palette.white};
            border-color: ${textOnly ? 'transparent' : `${palette.primary}`};
            color: ${palette.primary};

            &:hover {
                background-color: ${palette.secondary};
                border-color: ${palette.secondary};
                color: ${palette.white};
            }
        `
        : '';

export const generateDisabledStyling = ({ disabled }) =>
    disabled
        ? `
            border-color: ${palette.midGray} !important;
            cursor: not-allowed;
            pointer-events: none;

            &,
            span {
                color: ${palette.midGray} !important;
            }
        `
        : '';

export const generateShapeStyling = ({ shape }) => (shape === 'circle' ? 'border-radius: 50%;' : 'border-radius: 3px;');

export const PopupMenuStyledWrapper = styled.div`
    background: transparent;
    cursor: default;
    position: absolute;
`;

export const StyledWrapper = styled(
    React.forwardRef(({ full, ghost, iconAfterText, onlyIcon, popupMenuPosition, shape, textOnly, ...rest }, ref) => (
        <button ref={ref} {...rest} />
    )),
)`
    position: relative;
    border: 1px solid;
    display: inline-block;
    outline: 0 !important;
    font-size: ${typography.fontSm};
    line-height: ${typography.lineHeightSm};
    touch-action: manipulation;
    cursor: pointer;
    text-align: center;
    text-decoration: none;
    text-transform: uppercase;
    white-space: nowrap;
    padding: 8px 15px; // makes the height 40px
    transition: background-color 250ms ease-out, border-color 250ms ease-out, color 250ms ease-out;

    ${({ iconAfterText }) => (iconAfterText ? 'white-space: initial;' : '')}

    // PopupMenu styling:
    ${PopupMenuStyledWrapper} {
        ${generatePopupPositionBasedStyling}
    }

    // SVG styling:
    ${generateSVGStyling}

    .Button__tooltip-text {
        opacity: 0;
        position: absolute;
        transform: translateX(0);
        transition: opacity 250ms ease-out, transform 250ms ease-out;
        z-index: -1;
    }

    &:hover {
        .Button__tooltip-text {
            opacity: 1;
            transform: translateX(20px);
            transition: opacity 250ms ease-out, transform 250ms ease-out;
            transition-delay: 500ms;
            z-index: 0;
        }
    }

    // If buttons follow each other in a row
    // set a standardized whitespace between them
    & ~ & {
        margin-left: 9px;
    }

    ${generateFullPropsStyling}
    ${generateDisabledStyling}
    ${generateShapeStyling}
    ${generateGhostStyling}
`;
