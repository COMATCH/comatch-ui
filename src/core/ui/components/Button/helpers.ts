import { capitalizeFirstLetter, capitalizeAll } from '../../../utils/capitalizer';

function getText(text: string, textStyle: string) {
    switch (textStyle) {
        case 'capitalizeFirstLetter':
            return capitalizeFirstLetter(text);
        case 'capitalizeAll':
            return capitalizeAll(text);
        default:
            return text;
    }
}

export { getText };
