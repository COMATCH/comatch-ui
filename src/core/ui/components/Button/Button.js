import React from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import classNames from 'classnames';

import { Badge } from '../Badge';
import { StyledWrapper } from './StyledWrapper';

const propTypes = {
    /**
     * Additional class names
     */
    className: PropTypes.string,
    disabled: PropTypes.bool,
    /**
     * Alternative styling with white background and primary colored text
     */
    ghost: PropTypes.bool,
    /**
     * If true, uses `<a>` instead of `<button>`
     */
    href: PropTypes.string,
    /**
     * HTML id attribute
     */
    id: PropTypes.string,
    /**
     * SVG icon
     */
    icon: PropTypes.node,
    iconAfterText: PropTypes.bool,
    onClick: PropTypes.func,
    /**
     * To be used with `href`, could set for example `target="_blank"`
     */
    target: PropTypes.string,
    /**
     * Text content of `Button`
     */
    text: PropTypes.string,
    /**
     * Text content of `Button`'s tooltip
     */
    tooltipText: PropTypes.string,
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
};

const defaultProps = {
    className: '',
    disabled: false,
    ghost: false,
    href: '',
    id: null,
    icon: null,
    iconAfterText: false,
    onClick: noop,
    target: null,
    text: '',
    tooltipText: '',
    type: 'button',
};

/**
 * A `Button` component with different stylistic variations. By default the styling is `full`,
 * but `ghost` is an alternative version with reversed colors.
 * 
 * If `href` is set an `<a>` is used instead of `<button>`. If the `target` prop is set to `_blank`
 * in combination with `href`, then it will also add the prop `rel="noopener noreferrer"` for
 * additional security.
 */

export const Button = ({
    id,
    className,
    href,
    target,
    onClick,

    ghost,
    disabled,

    text,
    tooltipText,
    type,
    icon,
    iconAfterText,
}) => {
    const styledProps = {
        disabled,
        ghost,
        full: !ghost,
        onlyIcon: !text,
        iconAfterText: text && iconAfterText,

        ...(href && { as: 'a' }),
    };
    const content = iconAfterText ? (
        <span>
            {text}
            {icon}
        </span>
    ) : (
        <span>
            {icon}
            {text}
        </span>
    );

    const calculatedProps = {
        id,
        onClick,
        className: classNames('Button', className, {
            full: !ghost,
            disabled,
            ghost,
            'only-icon': !text,
            'icon-after-text': text && iconAfterText,
        }),

        ...(href ? {
            href,
            target,
            rel: target && 'noopener noreferrer',
            ...(disabled && {
                onClick: (evt) => evt.preventDefault(),
            }),
        } : {
            type,
        }),
    };

    return (
        <StyledWrapper {...styledProps} {...calculatedProps}>
            {content}
            {
                !!tooltipText
                && (
                    <Badge className="Button__tooltip-text" text={tooltipText} size="xs" color="darkGray" />
                )
            }
        </StyledWrapper>
    );
};

Button.propTypes = propTypes;

Button.defaultProps = defaultProps;
