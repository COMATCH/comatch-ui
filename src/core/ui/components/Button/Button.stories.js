import React from 'react';
import Highlight from 'react-highlight';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons/faTimes';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Button } from './Button';

storiesOf('Button', module)
    .add('with primary style', () => (
        <>
            <Button text="Hello world" />
            <Highlight className="html">{'<Button text="Hello world" />'}</Highlight>
        </>
    ))
    .add('disabled with primary style', () => (
        <>
            <Button disabled text="Hello world" />
            <Highlight className="html">{'<Button disabled text="Hello world" />'}</Highlight>
        </>
    ))
    .add('with primary style and onClick handler', () => (
        <>
            <Button text="Hello world" onClick={action('clicked')} />
            <Highlight className="html">{`<Button text="Hello world" onClick={(event) => {}} />`}</Highlight>
        </>
    ))
    .add('with ghost style', () => (
        <>
            <Button ghost text="Hello world" />
            <Highlight className="html">{'<Button ghost text="Hello world" />'}</Highlight>
        </>
    ))
    .add('disabled with ghost style', () => (
        <>
            <Button ghost disabled text="Hello world" />
            <Highlight className="html">{'<Button ghost disabled text="Hello world" />'}</Highlight>
        </>
    ))
    .add('with icon', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} text="Hello world" />
            <Highlight className="html">
                {'<Button icon={<FontAwesomeIcon icon={faTimes} />} text="Hello world" />'}
            </Highlight>
        </>
    ))
    .add('with icon and ghost style', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} ghost={true} text="Hello world" />
            <Highlight className="html">
                {'<Button icon={<FontAwesomeIcon icon={faTimes} />} ghost={true} text="Hello world" />'}
            </Highlight>
        </>
    ))
    .add('with icon on the right', () => (
        <>
            <Button iconAfterText={true} icon={<FontAwesomeIcon icon={faTimes} />} text="Hello world" />
            <Highlight>
                {'<Button iconAfterText={true} icon={<FontAwesomeIcon icon={faTimes} />} text="Hello world" />'}
            </Highlight>
        </>
    ))
    .add('with icon only', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} />
            <Highlight className="html">{'<Button icon={<FontAwesomeIcon icon={faTimes} />} />'}</Highlight>
        </>
    ))
    .add('with icon only and ghost style', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} ghost={true} />
            <Highlight className="html">
                {'<Button icon={<FontAwesomeIcon icon={faTimes} />} ghost={true} />'}
            </Highlight>
        </>
    ))

    .add('with shape set to circle', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} shape='circle' />
            <Highlight className="html">
                {'<Button icon={<FontAwesomeIcon icon={faTimes} />} shape="circle" />'}
            </Highlight>
        </>
    ))

    .add('with textOnly style', () => (
        <>
            <Button icon={<FontAwesomeIcon icon={faTimes} />} textOnly />
            <Highlight className="html">
                {'<Button icon={<FontAwesomeIcon icon={faTimes} />} textOnly />'}
            </Highlight>
        </>
    ))
    .add('with tooltipText', () => (
        <>
            <Button text="Hello world" tooltipText="Tooltip Text" />
            <Highlight className="html">
                {'<Button text="Hello world" tooltipText="Tooltip Text" />'}
            </Highlight>
        </>
    ));
