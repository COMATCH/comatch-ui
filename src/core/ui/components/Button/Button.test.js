import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import noop from 'lodash/noop';
import { Button } from './Button';
import '../../../config/tests/setup';

describe('Button', () => {
    it('should render the default button correctly', () => {
        const button = renderer.create(<Button />).toJSON();
        expect(button).toMatchSnapshot();
    });

    it('should render the ghost button correctly', () => {
        const button = renderer.create(<Button ghost={true} />).toJSON();
        expect(button).toMatchSnapshot();
    });

    it('should render the disabled button correctly', () => {
        const button = renderer.create(<Button disabled={true} />).toJSON();
        expect(button).toMatchSnapshot();
    });

    it('should render the ghost and disabled button correctly', () => {
        const button = renderer.create(<Button ghost={true} disabled={true} />).toJSON();
        expect(button).toMatchSnapshot();
    });
    it('should render the circle shaped button correctly', () => {
        const button = renderer.create(<Button shape='circle' />).toJSON();
        expect(button).toMatchSnapshot();
    });

    it('should render the textOnly and ghost button correctly', () => {
        const button = renderer.create(<Button textOnly />).toJSON();
        expect(button).toMatchSnapshot();
    });

    it('should have the correct default props', () => {
        const button = mount(<Button />);
        expect(button.prop('className')).toEqual('');
        expect(button.prop('disabled')).toEqual(false);
        expect(button.prop('ghost')).toEqual(false);
        expect(button.prop('href')).toEqual('');
        expect(button.prop('id')).toEqual(null);
        expect(button.prop('icon')).toEqual(null);
        expect(button.prop('iconAfterText')).toEqual(false);
        expect(button.prop('onClick')).toEqual(noop);
        expect(button.prop('target')).toEqual(null);
        expect(button.prop('text')).toEqual('');
        expect(button.prop('type')).toEqual('button');
        expect(button.prop('shape')).toEqual('rectangle');
    });

    it('should have the correct passed props', () => {
        const mockFunction = jest.fn();
        const props = {
            className: 'test className',
            disabled: true,
            ghost: true,
            href: 'test href',
            id: 'test id',
            onClick: mockFunction,
            target: 'test target',
            text: 'test text',
            textOnly: true,
            type: 'submit',
            shape: 'circle',
        };

        const button = mount(<Button {...props} />);
        expect(button.prop('className')).toEqual('test className');
        expect(button.prop('disabled')).toEqual(true);
        expect(button.prop('ghost')).toEqual(true);
        expect(button.prop('href')).toEqual('test href');
        expect(button.prop('id')).toEqual('test id');
        expect(button.prop('onClick')).toEqual(mockFunction);
        expect(button.prop('target')).toEqual('test target');
        expect(button.prop('text')).toEqual('test text');
        expect(button.prop('textOnly')).toEqual(true);
        expect(button.prop('type')).toEqual('submit');
        expect(button.prop('shape')).toEqual('circle');
    });

    it('should render the text', () => {
        const props = {
            text: 'text',
        };
        const button = mount(<Button {...props} />);
        expect(button.find('span').text()).toEqual('text');
    });

    it('should render the icon', () => {
        const props = {
            icon: 'icon',
        };
        const button = mount(<Button {...props} />);
        expect(button.find('span').text()).toEqual('icon');
    });

    describe('when iconAfterText prop is default', () => {
        it('should render first the icon and then the text', () => {
            const props = {
                text: 'text',
                icon: 'icon',
            };
            const button = mount(<Button {...props} />);
            expect(button.find('span').text()).toEqual('icontext');
        });
    });

    describe('when iconAfterText prop is false', () => {
        it('should render first the icon and then the text', () => {
            const props = {
                icon: 'icon',
                iconAfterText: false,
                text: 'text',
            };
            const button = mount(<Button {...props} />);
            expect(button.find('span').text()).toEqual('icontext');
        });
    });

    describe('when iconAfterText prop is true', () => {
        it('should render first the text and then the icon', () => {
            const props = {
                icon: 'icon',
                iconAfterText: true,
                text: 'text',
            };
            const button = mount(<Button {...props} />);
            expect(button.find('span').text()).toEqual('texticon');
        });
    });

    describe('when iconAfterText prop is true', () => {
        it('should render first the text and then the icon', () => {
            const props = {
                icon: 'icon',
                iconAfterText: true,
                text: 'text',
            };
            const button = mount(<Button {...props} />);
            expect(button.find('span').text()).toEqual('texticon');
        });
    });

    describe('when href prop is empty string', () => {
        it('should render a button element', () => {
            const button = mount(<Button />);
            expect(button.find('button').exists()).toBe(true);
        });

        describe('and disabled prop is false', () => {
            describe('and the button is clicked', () => {
                it('should run the passed onClick function', () => {
                    const mockFunction = jest.fn();
                    const button = mount(<Button onClick={mockFunction} />);
                    button.find('button').simulate('click');
                    expect(mockFunction.mock.calls.length).toEqual(1);
                });
            });
        });

        describe('and disabled prop is true', () => {
            describe('and the button is clicked', () => {
                it('should not run the passed onClick function', () => {
                    const mockFunction = jest.fn();
                    const button = mount(<Button disabled={true} onClick={mockFunction} />);
                    button.find('button').simulate('click');
                    expect(mockFunction.mock.calls.length).toEqual(0);
                });
            });
        });
    });

    describe('when href prop is a string other than an empty string', () => {
        it('should render an anchor element', () => {
            const props = {
                href: 'test href',
                iconAfterText: true,
            };
            const button = mount(<Button {...props} />);
            expect(button.find('a').exists()).toBe(true);
        });

        describe('and disabled prop is false', () => {
            describe('and the button is clicked', () => {
                it('should run the passed onClick function', () => {
                    const mockFunction = jest.fn();
                    const button = mount(<Button onClick={mockFunction} />);
                    button.find('button').simulate('click');
                    expect(mockFunction.mock.calls.length).toEqual(1);
                });
            });
        });

        describe('and disabled prop is true', () => {
            describe('and the button is clicked', () => {
                it('should not run the passed onClick function', () => {
                    const mockFunction = jest.fn();
                    const button = mount(<Button disabled={true} onClick={mockFunction} />);
                    button.find('button').simulate('click');
                    expect(mockFunction.mock.calls.length).toEqual(0);
                });
            });
        });
    });
});
