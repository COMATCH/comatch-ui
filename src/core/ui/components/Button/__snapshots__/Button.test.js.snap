// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render the button with tooltipText correctly 1`] = `
<button
  className="Button full only-icon sc-bwzfXH dLJvGH"
  disabled={false}
  id={null}
  onClick={[Function]}
  type="button"
>
  <span>
    
  </span>
  <div
    className="Badge Button__tooltip-text darkGray xs sc-bdVaJa icLWVk"
    color="darkGray"
    size="xs"
  >
    Testing tooltip text
  </div>
</button>
`;

exports[`Button should render the default button correctly 1`] = `
<button
  className="Button full only-icon sc-bwzfXH dLJvGH"
  disabled={false}
  id={null}
  onClick={[Function]}
  type="button"
>
  <span>
    
  </span>
</button>
`;

exports[`Button should render the disabled button correctly 1`] = `
<button
  className="Button full disabled only-icon sc-bwzfXH fvfiOE"
  disabled={true}
  id={null}
  onClick={[Function]}
  type="button"
>
  <span>
    
  </span>
</button>
`;

exports[`Button should render the ghost and disabled button correctly 1`] = `
<button
  className="Button disabled ghost only-icon sc-bwzfXH cCkvGP"
  disabled={true}
  id={null}
  onClick={[Function]}
  type="button"
>
  <span>
    
  </span>
</button>
`;

exports[`Button should render the ghost button correctly 1`] = `
<button
  className="Button ghost only-icon sc-bwzfXH jaJXmj"
  disabled={false}
  id={null}
  onClick={[Function]}
  type="button"
>
  <span>
    
  </span>
</button>
`;
